web/ts/TUMLiveVjs.ts

```diff
@@ -121,6 +118,75 @@ const skipSilence = function (options) {
     });
 };
 
+/**
+ * @function watchProgress
+ * Saves and retrieves the watch progress of the user as a fraction of the total watch time
+ * @param streamID The ID of the currently watched stream
+ * @param lastProgress The last progress fetched from the database
+ */
+const watchProgress = function (streamID: number, lastProgress: float64) {
+    this.ready(() => {
+        let duration;
+        let timer;
+        let initialized = false;
+        let interval = 10000;
+
+        // Fetch the user's stream progress from the database and set the time on load
+        this.on('loadedmetadata', () => {
+            duration = this.duration();
+            setProgress();
+        });
+
+        // iPhone/iPad need to play the video first, so they depend on a different event
+        // More info: https://www.w3.org/TR/html5/embedded-content-0.html#mediaevents
+        this.on('canplaythrough', () => {
+            duration = this.duration();
+            setProgress();
+        });
+
+        const setProgress = () => {
+            if (!initialized) {
+                this.currentTime(lastProgress * duration);
+            }
+            initialized = true;
+        }
+
+        const reportProgress = (currentTime) => {
+            const progress = currentTime / duration;
+            postData("/api/progressReport", {
+                "streamID": streamID,
+                "progress": progress
+            }).then(r => {
+                    if (r.status !== 200) {
+                        console.log(r);
+                        interval *= 2; // Binary exponential backoff for load balancing
+                    }
+                }
+            );
+        }
+
+        // Triggered when user presses play
+        this.on('play', () => {
+            timer = setInterval(() => { reportProgress(this.currentTime()) }, interval);
+        });
+
+        // Triggered on pause and skipping the video
+        this.on('pause', () => {
+            reportProgress(this.currentTime());
+            clearInterval(timer);
+        })
+
+        // Triggered when the video has no time left
+        this.on('ended', () => {
+            // Maybe we want to set a flag here in the future to mark the video as watched
+            reportProgress(0);
```

Comment:
Let's not reset the progress here, I'd like to keep it (even if it approaches 100%) to indicate that the video is watched through. This way we can mark them as viewed in the frontend or let users filter their lectures in the future (perhaps something for a follow-up?)

`setProgress()` would then need to check the progress and set it to 0 if it's close to the end (>95% maybe?)